tests/lora/test_lora_layers_flux.py

```diff
@@ -430,6 +430,97 @@ def test_correct_lora_configs_with_different_ranks(self):
+        # another lora with correct shapes is loaded. This is not supported at the moment and should raise an error.
+        # When we do support it, this test should be removed. Context: https://github.com/huggingface/diffusers/issues/10180
+        components, _, _ = self.get_dummy_components(FlowMatchEulerDiscreteScheduler)
+        pipe = self.pipeline_class(**components)
+        pipe = pipe.to(torch_device)
+        pipe.set_progress_bar_config(disable=None)
+
+        logger = logging.get_logger("diffusers.loaders.lora_pipeline")
+        logger.setLevel(logging.DEBUG)
+
+        out_features, in_features = pipe.transformer.x_embedder.weight.shape
+        rank = 4
+
+        shape_expander_lora_A = torch.nn.Linear(2 * in_features, rank, bias=False)
+        shape_expander_lora_B = torch.nn.Linear(rank, out_features, bias=False)
+        lora_state_dict = {
+            "transformer.x_embedder.lora_A.weight": shape_expander_lora_A.weight,
+            "transformer.x_embedder.lora_B.weight": shape_expander_lora_B.weight,
+        }
+        with CaptureLogger(logger) as cap_logger:
+            pipe.load_lora_weights(lora_state_dict, "adapter-1")
+            self.assertTrue(check_if_lora_correctly_set(pipe.transformer), "Lora not correctly set in denoiser")
+
+        self.assertTrue(pipe.transformer.x_embedder.weight.data.shape[1] == 2 * in_features)
+        self.assertTrue(pipe.transformer.config.in_channels == 2 * in_features)
+        self.assertTrue(cap_logger.out.startswith("Expanding the nn.Linear input/output features for module"))
+
+        normal_lora_A = torch.nn.Linear(in_features, rank, bias=False)
+        normal_lora_B = torch.nn.Linear(rank, out_features, bias=False)
+        lora_state_dict = {
+            "transformer.x_embedder.lora_A.weight": normal_lora_A.weight,
+            "transformer.x_embedder.lora_B.weight": normal_lora_B.weight,
+        }
+
+        # The first lora expanded the input features of x_embedder. Here, we are trying to load a lora with the correct
+        # input features before expansion. This should raise an error about the weight shapes being incompatible.
+        self.assertRaisesRegex(
+            RuntimeError,
+            "size mismatch for x_embedder.lora_A.adapter-2.weight",
+            pipe.load_lora_weights,
+            lora_state_dict,
+            "adapter-2",
+        )
+
+        # Test the opposite case where the first lora has the correct input features and the second lora has expanded input features.
+        # This should raise a runtime error on input shapes being incompatible. But it doesn't. This is because PEFT renames the
+        # original layers as `base_layer` and the lora layers with the adapter names. This makes our logic to check if a lora
+        # weight is compatible with the current model incorrect. This should be addressed when attempting support for
+        # https://github.com/huggingface/diffusers/issues/10180 (TODO)
```

Comment:
Could you provide some concrete LoCs as references for what you mean by:

> This makes our logic to check if a lora weight is compatible with the current model incorrect.

Would also love to understand how this relates to how `peft` names things.

Reply:
The lines in question are: https://github.com/huggingface/diffusers/blob/d041dd504058ac6b0fde3eb767eb6844d8d577b8/src/diffusers/loaders/lora_pipeline.py#L2319-L2322

When the first lora layer is loaded, assuming it is named `adapter-1` and the layer in question is `x_embedder`, the nn.Linear layer names are something like [`x_embedder`]. [This](https://github.com/huggingface/diffusers/blob/d041dd504058ac6b0fde3eb767eb6844d8d577b8/src/diffusers/loaders/lora_pipeline.py#L2321) check passes because `x_embedder.lora_A.weight` is indeed a key in the lora state dict.

After the first lora is loaded, peft updates the layer names to something like: [`x_embedder.base_layer`, `x_embedder.adapter-1.lora_A`, `x_embedder.adapter-1.lora_B`]. 

So, when the second lora is loaded, it tries to find `x_embedder.base_layer.lora_A.weight` in the lora state dict, which does not exist. It needs to instead search for `x_embedder.lora_A.weight`. But this won't happen because the model state dict contains x_embedder original layers in a renamed key.

Note that I don't recall the exact layer names, so it may differ when you test and I'm just giving an example. The rough idea is that the current logic only works for the loading the first lora correctly IF it is a normal lora, and the second lora being loaded is a shape expander lora.

But as discussed in DM, this was not an anticipated use case - we only wanted to make control lora work as expected so the shape mismatch error on weights instead of inputs is OK for now.

Reply:
Hmm got it. Thanks Aryan. Just noting this is enough for now.